File: agent-nodejs/controllers/agentEndpointController.js
const catchAsync = require("../utils/catchAsync");
const AppError = require("../utils/appError");
const axios = require("axios");
const endpointManager = require("../endpointsmanager");
const logger = require("../utils/logger");

//to check agent server is on
exports.getAgentStatus = catchAsync(async (req, res, next) => {
  res.status(200).json({
    success: true,
    message: "Connected",
  });
});

//this will fetch all agent endpointsto maintain internal list
exports.fetchAgentConfig = async () => {
  try {
    logger.debug(
      "fetching agent config from " +
        `${process.env.BACKEND_URL}/api/v1/endpoint/${process.env.AGENT_ID}`
    );
    const response = await axios({
      method: "Get",
      url: `${process.env.BACKEND_URL}/api/v1/endpoint/${process.env.AGENT_ID}`,
      headers: {
        AGENT_ID: process.env.AGENT_ID,
        AGENT_TOKEN: process.env.AGENT_TOKEN,
      },
    });

    if (response.data.success) {
      logger.debug(JSON.stringify(response.data));
      return {
        success: true,
        data: response.data.data.doc.map((item) => {
          if (!item._id) return item;
          return {
            ...item,
            id: item._id,
          };
        }),
      };
    } else {
      logger.debug("Error in fetching agent config");
      logger.debug(JSON.stringify(response));
      return {
        success: false,
        message: "Something went wrong",
      };
    }
  } catch (err) {
    logger.debug("Error in fetching agent config");
    return {
      success: false,
      message: "Something went wrong",
    };
  }
};

//this will change agent endpoint status
exports.updateEndPointStatus = catchAsync(async (req, res, next) => {
  const id = req.params.endpointId;
  const authToken = req.body.authToken;

  let endpointResponse = await endpointManager.changeEndpointsStatus(
    id,
    authToken
  );

  if (!endpointResponse.success) {
    return next(new AppError(endpointResponse.error, 422));
  }

  const newEndPointDoc = endpointResponse.data.find((x) => x._id === id);
  res.status(200).json({
    success: true,
    data: {
      doc: newEndPointDoc,
    },
  });
});

//This will give status of endpoints
exports.getEndPointStatus = catchAsync(async (req, res, next) => {
  res.status(200).json({
    success: true,
    data: {
      doc: endpointManager.getEndpoints(),
    },
  });
});

exports.addEndpoint = catchAsync(async (req, res, next) => {
  let body = req.body;
  if (body._id) body.id = body._id;
  let endpointResponse = endpointManager.addEndpoint(body);
  res.status(200).json({
    success: true,
    data: {
      doc: endpointResponse,
    },
  });
});

exports.updateEndpoint = catchAsync(async (req, res, next) => {
  const id = req.params.endpointId;
  const endpoint = {
    name: req.body.name,
    endpointYaml: req.body.endpointYaml,
  };
  let endpointResponse = await endpointManager.updateEndpoint(id, endpoint);

  res.status(200).json({
    success: true,
    data: {
      doc: endpointResponse,
    },
  });
});

exports.deleteEndpoint = catchAsync(async (req, res, next) => {
  const id = req.params.endpointId;

  let endpointResponse = await endpointManager.deleteEndpoint(id);

  res.status(200).json({
    success: true,
    data: {
      doc: endpointResponse,
    },
  });
});

File: agent-python/controllers/agentEndpointController.py
import logging
import requests
import os
from utils.appError import AppError

from fastapi import HTTPException
from utils.logger import logger
import endpointsmanager     

def getAgentStatus():
    response= {"success": True, "message": "Connected"}
    logger.debug(response)
    return response
async def fetchAgentConfig():
    #  not user:
    #     raise HTTPException(
    #         status_code=404,
    #         detail="The user with this email does not exist in the system.",
    #     )
    try:
        logger.debug(f"fetching agent config from {os.environ['BACKEND_URL']}/api/v1/endpoint/{os.environ['AGENT_ID']}")
        response =  requests.get(
            f"{os.environ['BACKEND_URL']}/api/v1/endpoint/{os.environ['AGENT_ID']}",
            headers={"AGENT_TOKEN": os.environ['AGENT_TOKEN'],"AGENT_ID":os.environ['AGENT_ID']}
        )
        response= response.json()
        if response.get("success")==True:
            data= response.get("data")
            logger.debug(data)
            return {
                "success": True,
                "data": [
                    {**item, "id": item["_id"]} if item.get("_id") else item 
                    for item in data.get("doc")
                ]
            }
        else:
            logger.debug("Error in fetching agent config")
            logger.debug(response)
            return {"success": False, "message": "Something went wrong"}
    except Exception as err:
        logger.debug("Error in fetching agent config",err)
        return {"success": False, "message": "Something went wrong"}

async def updateEndPointStatus(endpointId: str,authToken:str):
  
    endpoint_response = await endpointsmanager.changeEndpointsStatus(endpointId,authToken)
    logger.debug(endpoint_response)
    if not endpoint_response.get("success")==True:
        raise AppError(status_code=404,detail=endpoint_response.get("error") or "Agent endpoint not updated")

    new_end_point_doc = next((x for x in endpoint_response.get("endpoints") if x["_id"] == endpointId), None)
    return {"success": True, "data": {"doc": new_end_point_doc}}

async def getEndPointStatus():
    response= {"success": True, "data": {"doc": endpointsmanager.getEndpoints()}}
    logger.debug(response)
    return response

async def addEndpoint(endpoint: dict):
    if "_id" in endpoint:
        endpoint["id"] = endpoint["_id"]
    endpoint_response = endpointsmanager.addEndpoint(endpoint)
    response= {"success": True, "data": {"doc": endpoint_response}}
    logger.debug(response)
    return response


async def updateEndpoint(endpointId: str,endpoint: dict):
    endpoint_response = await endpointsmanager.updateEndpoint(endpointId, endpoint)
    if not endpoint_response:
        return {"success": False, "error": "Endpoint not found"}
    response= {"success": True, "data": {"doc": endpoint_response}}
    logger.debug(response)
    return response

async def deleteEndpoint(endpointId: str):
    endpoint_response = endpointsmanager.deleteEndpoint(endpointId)
    response= {"success": True, "data": {"doc": endpoint_response}}
    logger.debug(response)
    return response 
